src/Distributions/UniformPowerOfTwo/Model.dfy

```diff
@@ -32,51 +34,125 @@ module UniformPowerOfTwoModel {
       Monad.Bind(Sample(n/2), UnifStep)
   }
 
-  lemma {:axiom} SampleTerminates(n: nat)
-    requires n > 0
-    ensures
-      var b := Sample(n - 1);
-      var c := (x: nat) => x < n;
-      && Independence.IsIndepFn(b)
-      && Quantifier.ExistsStar(WhileAndUntil.Helper2(b, c))
-      && WhileAndUntil.ProbUntilTerminates(b, c)
-
-  function SampleAlternative(n: nat, k: nat := 1, u: nat := 0): Monad.Hurd<nat>
-    requires k >= 1
-    decreases 2*n - k
+  // A tail recursive version of Sample, closer to the imperative implementation
+  function SampleTailRecursive(n: nat, u: nat := 0): Monad.Hurd<nat>
+    decreases n
   {
     (s: RandomNumberGenerator.RNG) =>
-      if k > n then
+      if n == 0 then
         (u, s)
       else
-        SampleAlternative(n, 2*k, if Monad.Head(s) then 2*u + 1 else 2*u)(Monad.Tail(s))
+        SampleTailRecursive(n / 2, if Monad.Head(s) then 2*u + 1 else 2*u)(Monad.Tail(s))
   }
 
-  // incomplete
+  // Equivalence of Sample and its tail-recursive version
   lemma SampleCorrespondence(n: nat, s: RandomNumberGenerator.RNG)
-    ensures SampleAlternative(n)(s) == Sample(n)(s)
+    ensures SampleTailRecursive(n)(s) == Sample(n)(s)
   {
-    var f := (m: nat) =>
-        var g := (b: bool) =>
-                   Monad.Return(if b then 2*m + 1 else 2*m);
-        Monad.Bind(Monad.Deconstruct, g);
     if n == 0 {
-    } else if n == 1 {
-      assert n / 2 == 0;
-      assert (0, s) == Sample(n/2)(s);
+      assert SampleTailRecursive(n)(s) == Sample(n)(s);
+    } else {
+      var k := Helper.Log2(n);
+      assert Power2(k - 1) <= n < Power2(k) by { Power2OfLog2(n); }
       calc {
+        SampleTailRecursive(n)(s);
+        { SampleTailRecursiveEqualIfSameLog2Floor(n, Power2(k - 1), k, 0, s); }
+        SampleTailRecursive(Power2(k - 1))(s);
+        Monad.Bind(Sample(Power2(-1)), (u: nat) => SampleTailRecursive(Power2(k - 1), u))(s);
+        { RelateWithTailRecursive(k - 1, k, s); }
+        Sample(Power2(k - 1))(s);
+        { SampleEqualIfSameLog2Floor(n, Power2(k - 1), k, s); }
         Sample(n)(s);
-        Monad.Bind(Sample(n/2), f)(s);
-        f(0)(s);
-        Monad.Bind(Monad.Deconstruct, (b: bool) => Monad.Return(if b then 1 else 0))(s);
-        Monad.Return(if Monad.Head(s) then 1 else 0)(Monad.Tail(s));
-        (if Monad.Head(s) then 1 else 0, Monad.Tail(s));
-        SampleAlternative(1, 2, if Monad.Head(s) then 1 else 0)(Monad.Tail(s));
-        SampleAlternative(1, 1, 0)(s);
-        SampleAlternative(n)(s);
       }
+    }
+  }
+
+  // A version of Power(2, k) extended to negative k's.
+  function Power2(k: int): nat {
+    if k < 0 then 0 else if k == 0 then 1 else Power2(k - 1) * 2
+  }
+
+  lemma Power2Greater1(k: nat)
+    ensures Power2(k) >= 1
+  {}
+
+  lemma Power2OfLog2(n: nat)
```

Comment:
Move these lemmas to `Helper.dfy`. Within the latter, remove the lemma `Log2BothSides` (maybe more) and substitute calls to it with calls to `Power2OfLog2`.